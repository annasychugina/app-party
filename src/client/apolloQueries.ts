import gql from 'graphql-tag';

export const GET_CHARACTERS = gql`
  query CharactersQuery($filter: FilterCharacter) {
    characters(filter: $filter) {
      results {
        id
        name
        image
      }
    }
  }
`;
