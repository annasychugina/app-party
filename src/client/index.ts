import ApolloClient, { InMemoryCache } from 'apollo-boost';
import gql from 'graphql-tag';

export const typeDefs = gql`
  type Query {
    charactersOnParty: [Character]!
    party: Party!
  }
  type Party {
    rick: Character!
    morty: Character!
  }
`;

const RICK_MORTY_API = 'https://rickandmortyapi.com/graphql';
const cache = new InMemoryCache({});

export const apolloClient = new ApolloClient({
  uri: RICK_MORTY_API,
  cache,
  typeDefs
});
