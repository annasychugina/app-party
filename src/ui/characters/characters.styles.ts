import styled from 'styled-components';

export const Grid = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin-left: -10px;
  margin-right: -10px;
`;

export const Column = styled.li`
  padding: 10px;
  width: 100%;
  height: 240px;
  max-width: 200px;
`;

export const ErrorText = styled.div`
  color: #ff0000;
  text-align: center;
`;

export const WarningText = styled.div`
  color: #dadada;
  text-align: center;
`;
