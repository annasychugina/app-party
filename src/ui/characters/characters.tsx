import React, {useCallback, useMemo, useEffect} from 'react';
import {Preloader} from '../preloader/preloader';
import {Card} from '../card/card';
import {useMutation, useLazyQuery} from '@apollo/react-hooks';
import {GET_CHARACTERS_QUERY, UPDATE_PARTY_CHARACTER} from '../../client/apolloQueries';
import resources from './config.json';
import {QueryCharactersArgs} from '../../types/graphql';
import {ICharacter} from '../../types/types';
import {Column, ErrorText, Grid, WarningText} from './characters.styles';
import {MIN_SYMBOLS_COUNT} from '../../constants';

interface IProps {
  searchTerm: string;
  debouncedSearchTerm: string;
  onChange: (value: string) => void;
  onRemoveCharacter: (id: string) => void;
  removedCharacters: string[];
}

export interface ICharactersData {
  results: ICharacter[] | null;
}

export interface ICharactersQuery {
  characters: ICharactersData | null;
}

export const Characters: React.FC<IProps> = ({debouncedSearchTerm, onRemoveCharacter, removedCharacters}: IProps) => {
  const [getCharacters, {loading, data, error}] = useLazyQuery<ICharactersQuery, QueryCharactersArgs>(
    GET_CHARACTERS_QUERY,
    {fetchPolicy: 'network-only'},
  );
  const [updatePartyCharacter] = useMutation(UPDATE_PARTY_CHARACTER);
  const characters = data?.characters?.results;
  const filteredCharacters = useMemo(
    () => characters && characters.filter(({id}: ICharacter) => removedCharacters.indexOf(id) === -1),
    [characters, removedCharacters],
  );

  useEffect(() => {
    if (debouncedSearchTerm.length > MIN_SYMBOLS_COUNT) {
      getCharacters({
        variables: {
          filter: {
            name: debouncedSearchTerm,
          },
        },
      });
    }
  }, [debouncedSearchTerm, getCharacters]);

  const handleClick = useCallback(
    (character: ICharacter) => {
      updatePartyCharacter({
        variables: {
          character,
        },
      });
    },
    [updatePartyCharacter],
  );

  if (loading) {
    return <Preloader />;
  }

  if (error) {
    return <ErrorText data-testid="error-text">{resources.errorText}</ErrorText>;
  }

  if (debouncedSearchTerm.length > MIN_SYMBOLS_COUNT && (!filteredCharacters || filteredCharacters.length === 0)) {
    return <WarningText data-testid="warning-text">{resources.warningText}</WarningText>;
  }

  return (
    <Grid>
      {filteredCharacters?.map(({id, image, name}) => (
        <Column key={id}>
          <Card
            imageUrl={image}
            onRemoveCharacter={(e: React.SyntheticEvent) => {
              e.stopPropagation();
              onRemoveCharacter(id);
            }}
            onClick={() => handleClick({id, image, name})}
          />
        </Column>
      ))}
    </Grid>
  );
};
