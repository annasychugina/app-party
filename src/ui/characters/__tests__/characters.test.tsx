import React from 'react';
import {MockedProvider, wait} from '@apollo/react-testing';
import {GET_CHARACTERS_QUERY, UPDATE_PARTY_CHARACTER} from '../../../client/apolloQueries';
import {Characters} from '../characters';
import {fireEvent, render, waitForDomChange, act, cleanup} from '@testing-library/react';

const deleteCharacter = {id: '1', name: 'rick 123', image: 'image rick'};
const mocks = [
  {
    request: {
      query: GET_CHARACTERS_QUERY,
      variables: {filter: {name: 'rick'}},
    },
    result: {
      data: {
        __typename: 'Characters',
        characters: {
          __typename: 'Character',
          results: [
            {
              id: '1',
              name: 'rick 123',
              image: 'image rick',
              __typename: 'Character',
            },
            {
              id: '2',
              name: 'rick 1234',
              image: 'image rick45',
              __typename: 'Character',
            },
          ],
        },
      },
    },
  },
  {
    request: {
      query: UPDATE_PARTY_CHARACTER,
      variables: {character: {id: '1', name: 'rick 123', image: 'image rick'}},
    },
    result: {
      data: {deleteCharacter},
    },
  },
];

const mocks2 = [
  {
    request: {
      query: GET_CHARACTERS_QUERY,
      variables: {filter: {name: 'rick'}},
    },
    error: new Error('Error'),
  },
  {
    request: {
      query: UPDATE_PARTY_CHARACTER,
      variables: {data: {id: '1', name: 'rick 123', image: 'image rick'}},
    },
    result: {
      data: {deleteCharacter},
    },
  },
];

const mock3 = [
  {
    request: {
      query: GET_CHARACTERS_QUERY,
      variables: {filter: {name: 'test'}},
    },
    result: {
      data: {
        __typename: 'Characters',
        characters: {
          __typename: 'Character',
          results: [],
        },
      },
    },
  },
  {
    request: {
      query: UPDATE_PARTY_CHARACTER,
      variables: {data: {id: '1', name: 'rick 123', image: 'image rick'}},
    },
    result: {
      data: {deleteCharacter},
    },
  },
];

describe('Characters', () => {
  afterEach(cleanup);
  it('should render error', async () => {
    await act(async () => {
      const {findByTestId} = render(
        <MockedProvider resolvers={{}} mocks={mocks2}>
          <Characters
            searchTerm={'rick'}
            removedCharacters={[]}
            debouncedSearchTerm={'rick'}
            onRemoveCharacter={() => {}}
            onChange={() => {}}
          />
        </MockedProvider>,
      );

      await wait(500); // wait for response
      const error = await findByTestId('error-text');
      expect(error).toBeTruthy();
    });
  });

  it('should render warning if no results', async () => {
    await act(async () => {
      const {findByTestId} = render(
        <MockedProvider resolvers={{}} mocks={mock3}>
          <Characters
            searchTerm={'test'}
            removedCharacters={[]}
            debouncedSearchTerm={'test'}
            onRemoveCharacter={() => {}}
            onChange={() => {}}
          />
        </MockedProvider>,
      );

      await wait(500);
      const warningText = await findByTestId('warning-text');
      expect(warningText).toBeTruthy();
    });
  });

  it('remove card on click correctly', async () => {
    await act(async () => {
      const removeCharacterMock = jest.fn();
      const component = render(
        <MockedProvider resolvers={{}} mocks={mocks}>
          <Characters
            searchTerm={'rick'}
            removedCharacters={[]}
            debouncedSearchTerm={'rick'}
            onRemoveCharacter={removeCharacterMock}
            onChange={() => {}}
          />
        </MockedProvider>,
      );
      await wait(500);
      const nodes = await component.getAllByTestId('remove');
      expect(nodes.length).toBe(2);
      fireEvent(
        nodes[0],
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
      expect(removeCharacterMock).toHaveBeenNthCalledWith(1, '1');
    });
  });

  it('mutation should be called when click the button', async () => {
    await act(async () => {
      const component = render(
        <MockedProvider resolvers={{}} mocks={mocks}>
          <Characters
            searchTerm={'rick'}
            removedCharacters={[]}
            debouncedSearchTerm={'rick'}
            onRemoveCharacter={() => {}}
            onChange={() => {}}
          />
        </MockedProvider>,
      );
      await waitForDomChange();
      const nodes = await component.getAllByTestId('character-card');
      expect(nodes.length).toBe(2);
      fireEvent(
        nodes[0],
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
      await wait(500);
      const updateCharacterResult = mocks[1].result;
      expect(updateCharacterResult).toMatchSnapshot();
    });
  });
});
