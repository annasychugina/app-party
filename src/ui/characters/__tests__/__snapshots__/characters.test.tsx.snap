// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters mutation should be called when click the button 1`] = `
Object {
  "data": Object {
    "deleteCharacter": Object {
      "id": "1",
      "image": "image rick",
      "name": "rick 123",
    },
  },
}
`;
